### PR TITLE
NH-70492 -  Otelcol: collecting Lambda platform metrics via Telemetry API

### PR DESCRIPTION
**Pull request by jerrytfleung**
## Changes
- Added support to the following metrics in telemetry API receiver

faas.coldstarts | Counter | When we see phase = init from initReport, fass.coldstarts++
-- | -- | --
faas.errors | Counter | When we see status != success from report, faas.errors++
faas.invocations | Counter | When we receive report, faas.invocations++
faas.timeouts | Counter | When we see status == timeout from report, faas.timeouts++

- Fixed `semconv.AttributeFaaSMaxMemory`. It has to be in [bytes](https://opentelemetry.io/docs/specs/semconv/attributes-registry/faas/)
- Updated unit test cases

Confluence page: https://swicloud.atlassian.net/wiki/spaces/NIT/pages/4639653945/otelcol+collecting+Lambda+metrics

### CODE REVIEW

File: collector/internal/telemetryapi/types.go [hunk 1]
```diff
@@ -24,6 +24,25 @@ const (
 	PlatformInitStart EventType = Platform + ".initStart"
 	// PlatformInitRuntimeDone is used when function initialization ended.
 	PlatformInitRuntimeDone EventType = Platform + ".initRuntimeDone"
+	// PlatformInitReport is used to report of function initialization.
```
Comment (tammy-baylis-swi) [inline collector/internal/telemetryapi/types.go:27]: Should there also be an entry for `PlatformExtension` for what's in [collector/receiver/telemetryapireceiver/types.go](https://github.com/solarwinds/opentelemetry-lambda/pull/41/files#diff-64884b6646bfae2414b23b292c061d38b5bb5fb823f4414b79ed56aee96ecf81)? Or not necessary?

File: collector/receiver/telemetryapireceiver/receiver.go [hunk 2]
```diff
@@ -196,9 +216,106 @@ func (r *telemetryAPIReceiver) createTraces(slice []telemetryapi.Event) (ptrace.
...
+			metrics.Metadata().PutStr("type", el.Type)
+			metrics.SetName("faas.invocations")
+			sum := metrics.SetEmptySum()
+			sum.SetIsMonotonic(true)
+			sum.SetAggregationTemporality(pmetric.AggregationTemporalityCumulative)
+			dp := sum.DataPoints().AppendEmpty()
+			dp.SetIntValue(r.invocationsCounter)
+			dp.SetStartTimestamp(pcommon.NewTimestampFromTime(r.metricsStartTime))
+			dp.SetTimestamp(pcommon.NewTimestampFromTime(time.Now()))
+			dp.Attributes().PutStr(semconv.AttributeFaaSTrigger, semconv.AttributeFaaSTriggerOther)
```
Comment (tammy-baylis-swi) [inline collector/receiver/telemetryapireceiver/receiver.go]: There are some parts of each `case` and within the `if-else` below that repeat a bit, though not completely. Would it make sense to factor our some helper functions, or would context be lost and/or too un-readable?

**Review by tammy-baylis-swi — APPROVED** [review]
Lgtm, though would be good to get more eyes on this one. I just left questions.

File: collector/internal/telemetryapi/types.go (hunk 1 above)
Comment (jerrytfleung) [reply collector/internal/telemetryapi/types.go:27]: I was following the [telemetry API http schema](https://docs.aws.amazon.com/lambda/latest/dg/samples/events_http_schema_v2022_12_13.zip) to create all those type first, though we are not using it in this PR. Suggest to keep it there.

File: collector/receiver/telemetryapireceiver/receiver.go (hunk 2 above)
Comment (jerrytfleung) [reply collector/receiver/telemetryapireceiver/receiver.go]: Sure. Done

File: collector/internal/telemetryapi/types.go (hunk 1 above)
Comment (tammy-baylis-swi) [reply collector/internal/telemetryapi/types.go:27]: Sounds good!

**Pull request merged**